include/drivers/eeprom.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2019 Vestas Wind Systems A/S
+ *
+ * Heavily based on drivers/flash.h which is:
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2016 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for EEPROM drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_EEPROM_H_
+#define ZEPHYR_INCLUDE_DRIVERS_EEPROM_H_
+
+/**
+ * @brief EEPROM Interface
+ * @defgroup eeprom_interface EEPROM Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef int (*eeprom_api_read)(struct device *dev, off_t offset, void *data,
+			       size_t len);
+typedef int (*eeprom_api_write)(struct device *dev, off_t offset,
+				const void *data, size_t len);
+typedef int (*eeprom_api_write_protection)(struct device *dev, bool enable);
+
+struct eeprom_driver_api {
+	eeprom_api_read read;
+	eeprom_api_write write;
+	eeprom_api_write_protection write_protection;
+	const size_t size;
+};
+
+/**
+ *  @brief Read data from EEPROM
+ *
+ *  @param dev EEPROM device
+ *  @param offset Address offset to read from.
+ *  @param data Buffer to store read data.
+ *  @param len Number of bytes to read.
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_read(struct device *dev, off_t offset, void *data,
+			  size_t len);
+
+static inline int z_impl_eeprom_read(struct device *dev, off_t offset,
+				     void *data, size_t len)
+{
+	const struct eeprom_driver_api *api = dev->driver_api;
+
+	return api->read(dev, offset, data, len);
+}
+
+/**
+ *  @brief Write data to EEPROM
+ *
+ *  Prior to the invocation of this API, the eeprom_write_protection_set needs
+ *  to be called first to disable the write protection.
+ *
+ *  @param dev EEPROM device
+ *  @param offset Address offset to write data to.
+ *  @param data Buffer with data to write.
+ *  @param len Number of bytes to write.
+ *
+ *  @return 0 on success, negative errno code on failure.
+ */
+__syscall int eeprom_write(struct device *dev, off_t offset, const void *data,
```

Comment:
Why do you need a syscall here? As I understand the API, you will only use the i2c or SPI APIs, which are syscalls themselves.

Reply:
The two drivers, I have in progress, only use the I2C and/or SPI APIs, correct. But other EEPROM driver can be added at a later point (e.g. the NXP Kinetis series has the ability to expose part of its flash as EEPROM).